Begin adopting cardano-api throughout the Hydra code-base

KtorZ opened the pull request:
So far, I have:

  - Refactored the Hydra.Ledger to avoid name-conflicts.
  - Replaced our custom 'Balance' type with a 'ValueType' family on `IsTx'.
  - Changed '{To,From}CBOR' constaints with 'SerialiseAsCBOR' from cardano-api's for easier interop.
  - Used cardano-api's `Tx AlonzoEra` as the underlying transaction type for Hydra.Ledger.Cardano.
  - Moved orphan JSON instances from Hydra.Ledger.Cardano into Hydra.Ledger.Cardano.Orphans.
  - Reviewed existing instances to work with Alonzo instead of Mary.

  Still to do:

  - `fromLedgerTx` to be implemented to recover a cardano-api's Tx from.
    a cardano-ledger-specs's Tx.
  - Adapt all changes to the rest of the code-base!

KtorZ committed:
Begin adopting cardano-api throughout the Hydra code-base

  So far, I have:

  - Refactored the Hydra.Ledger to avoid name-conflicts.
  - Replaced our custom 'Balance' type with a 'ValueType' family on `IsTx'.
  - Changed '{To,From}CBOR' constaints with 'SerialiseAsCBOR' from cardano-api's for easier interop.
  - Used cardano-api's `Tx AlonzoEra` as the underlying transaction type for Hydra.Ledger.Cardano.
  - Moved orphan JSON instances from Hydra.Ledger.Cardano into Hydra.Ledger.Cardano.Orphans.
  - Reviewed existing instances to work with Alonzo instead of Mary.

  Still to do:

  - `fromLedgerTx` to be implemented to recover a cardano-api's Tx from.
    a cardano-ledger-specs's Tx.
  - Adapt all changes to the rest of the code-base!

KtorZ committed:
Keep working on absorbing cardano-api in the Hydra.Ledger.Cardano.

  Only remains to translate the generators now... plus the repercusions on the rest of the codebase.

KtorZ committed:
Finish translating generators to cardano-api.

KtorZ committed:
Revert using SerialiseAsCBOR instead of ToCBOR/FromCBOR

  This is too big of a change at the moment, with debatable benefits. We'll resort to the same trick as before with an orphan instance on Tx

KtorZ committed:
Adjust hydra-node modules to interface renaming (Tx => IsTx, Utxo => UtxoType...)

KtorZ committed:
Use Mary generator for Tx and UTxO, to keep test working for now.

KtorZ committed:
Add scripts and bootstrap witnesses to JSON Orphans.

ch1bo committed:
Apply HLint hints

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Tx.hs:
```diff
@@ -55,8 +55,7 @@ import Hydra.Data.Party (partyFromVerKey, partyToVerKey)
 import qualified Hydra.Data.Utxo as OnChain
-import Hydra.Ledger (Utxo)
-import Hydra.Ledger.Cardano (CardanoTx, utxoToJSON)
+import Hydra.Ledger.Cardano (CardanoTx, Utxo)
```
Should: Looking at these names, `CardanoUtxo` seems to be a better name (also given the many `UTxO` types already existing)

KtorZ replied:
Indeed, I was thinking the same yesterday for consistency. But, at the same time, we use TxId and Value from cardano-api. So, unless we also prefix them the same way, the rationale for CardanoUtxo is a bit diminished.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+mkVkAddress ::
+  VerificationKey PaymentKey ->
+  Address ShelleyAddr
+mkVkAddress vk =
```
There is an overlap with `CardanoClient.buildAddress` and I'd argue that we only ever need the `AddressInEra Era` with `cardano-api` (all usage sites currently do) so maybe using `shelleyAddressInEra` inside this function to get that type back.

Could: Refactor and DRY with `buildAddress` to ultimately provide a function `mkAddress :: VerificationKey PaymentKey -> NetworkId -> AddressInEra Era`

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+-- of 'Ledger.UTXO'and provide 'Monoid' and 'Foldable' instances to make utxo
+-- manipulation bareable.
+newtype Utxo' era out = Utxo
+  { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto era)) out
```
Why do we need two type variables? `newtype Utxo' era = Utxo { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto era)) (Ledger.Alonzo.TxOut era) }` should work as well?
Then, the question is. Do you don't seem to use `Utxo'` with varying types, so this one should suffice and simplify code here and there a bit:
`newtype Utxo = Utxo { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto LedgerEra)) (Ledger.Alonzo.TxOut LedgerEra) }`

ch1bo replied:
Ah.. I guess I found it. Because of the `Functor` instance? If yes and you want to keep the `Functor`, this requires a comment.

IMO the `Functor`of `Utxo'` is a bit a farce, as it is only used with `out = Ledger.Alonzo.TxOut LedgerEra` so `fmap`'s practical signature would rather be `fmap :: (Ledger.Alonzo.TxOut LedgerEra -> Ledger.Alonzo.TxOut LedgerEra) -> Utxo -> Utxo`. Which function I'd personally would name `mapOutputs` or give it an shorthand operator if it is often used. But this is quite opinionated :)

KtorZ replied:
It is mainly for the Foldable instance. The functor instance is a nice collateral which is handy for converting to and from ledger types.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
 
-type CardanoKeyPair = Cardano.KeyPair 'Cardano.Payment StandardCrypto
+instance Monoid Utxo where
+  mempty = Utxo mempty
```
Could: Not use explicit but `deriving newtype` instances for `Semigroup`and `Monoid`

KtorZ replied:
Hmmm. It wasn't possible before because I was wrapping `Ledger.UTxO`, but I later changed that so that's a good simplification indeed!

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
---
+prettyUtxo :: (TxIn, TxOut ctx era) -> Text
+prettyUtxo (k, TxOut _ (txOutValueToValue -> v) _) =
+  T.drop 54 (renderTxIn k) <> " ↦ " <> prettyValue v
```
Must: Can you add a comment why `54` is fine?

KtorZ replied:
Yes, though the function pre-existed this PR :sweat_smile:

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+-- FIXME: This function is wrong! It's mapping a transaction's own inputs to its
+-- own outputs. Whoops. It's currently used in Hydra.Chain.Direct.Tx where
+-- the calling code only look at the outputs of the transactions and also in the
+-- generators of the local-cluster (Whoops bis).
```
You are right. Would a correct implementation where the `TxId` and `TxIx` are correctly referring to the given `CardanoTx`s outputs be helpful to have in your opinion?

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+  txId = getTxId . getTxBody
+  balance (Utxo u) =
+    let aggregate (Ledger.Alonzo.TxOut _ value _) = (<>) (fromMaryValue value)
+     in Map.foldr aggregate mempty u
```
As you have a `Foldable` instance on `Utxo` (see comment above, but if we keep it, we should use it), you could use `foldMap` and leverage the `Monoid Value`. So this should work:

```suggestion
  balance = foldMap (\(Ledger.Alonzo.TxOut _ value _) -> fromMaryValue value)
```

KtorZ replied:
Indeed, I just took the code as it was, but can be simplified now :)

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
 
-instance Crypto crypto => ToJSON (Cardano.TxIn crypto) where
-  toJSON = toJSON . txInToText
+toLedgerAddr :: AddressInEra Era -> Ledger.Addr Ledger.StandardCrypto
```
Could: All these functions are already in the `cardano-api` right? e.g. https://github.com/input-output-hk/cardano-node/blob/2cbe363874d0261bc62f52185cf23ed492cf4859/cardano-api/src/Cardano/Api/Address.hs#L536 for this one. Can we flag them with some comment in the code base and request access to it from upstream? Not sure if it is wise, but re-doing this feels a bit pointless.

KtorZ replied:
Partly, most of them are, some aren't exactly in the same form. But yes, ideally I would prefer not having to re-define them.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
 
 --
--- Output
+-- Formatting
```
`prettyUtxo` would fit in here as well if you would like to sort them like this.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano.hs:
```diff
@@ -95,712 +82,510 @@ cardanoLedger =
+-- some of our tests / interfaces aren't fully ready for Alonzo-specific
+-- details. We later changed the ledger's internals to work with Alonzo-era
+-- specific types, and, to make this in incremental steps, this function still
+-- generates Mary transactions, but cast them to Alonzo's.
```
:+1:

ch1bo replied:
How are the passed `Utxo` utilized? Are all of them spent? or just some arbitrary ones of them?

KtorZ replied:
Most of them are spent.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Orphans.hs:
```diff
@@ -0,0 +1,507 @@
+-- | Orphans instances (mostly ToJSON/FromJSON) required by Hydra.Ledger.Cardano
+-- to satisfies our various internal interfaces.
+--
+-- Nothing to see here.
```
So all of these are candidates we were considering to use the `cardano-api` ones? Especially this note from `Cardano.Api.Orphans` sounds like they intend to have JSON instances for their wrapper types we could rely on (in the mid-term): https://github.com/input-output-hk/cardano-node/blob/2cbe363874d0261bc62f52185cf23ed492cf4859/cardano-api/src/Cardano/Api/Orphans.hs#L66-L67

KtorZ replied:
Actually, this module only contains instances not present in Cardano.Api (that is, most `FromJSON` ones and some `ToJSON` ones)

ch1bo commented on hydra-node/src/Hydra/Ledger/Simple.hs:
```diff
@@ -66,24 +69,30 @@ instance ToCBOR SimpleTx where
 
 -- |An identifier for a single output of a 'SimpleTx'.
-newtype TxIn = TxIn {unTxIn :: Integer}
+newtype MockTxIn = MockTxIn {unMockTxIn :: Integer}
```
We called this `SimpleTxIn` in the `ch1bo/commit-real-utxo` branch.
Could: Rename to `SimpleTxIn` or we change the other branch to `MockTxIn`. Any preferences on the name?

KtorZ replied:
`SimpleTxIn` feels more coherent.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/MockServer.hs:
```diff
@@ -192,7 +191,7 @@ mockChainSyncServer db =
 
   serverStIdle :: Int -> ServerStIdle Block (Point Block) (Tip Block) m ()
-  serverStIdle !cursor =
+  serverStIdle ! cursor =
```
Wtf fourmolu.

KtorZ replied:
Yeah..

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -106,7 +106,7 @@ spec =
          in len < maxTxSize
               & label (show (len `div` 1024) <> "kB")
-              & counterexample ("UTXO's size: " <> show (Map.size $ unUTxO utxo))
+              & counterexample ("UTXO's size: " <> show (length utxo))
```
Okay this is nice :heart_eyes:

ch1bo commented on hydra-node/test/Hydra/Ledger/CardanoSpec.hs:
```diff
@@ -9,21 +9,21 @@ import Hydra.Prelude
-  roundtripAndGoldenSpecs (Proxy @CardanoTxWitnesses)
+  -- FIXME: Roundtrip instances for all JSON types we depend on
+  roundtripAndGoldenSpecs (Proxy @Utxo)
+  -- roundtripAndGoldenSpecs (Proxy @CardanoTxWitnesses)
```
Commented code.

ch1bo commented on hydra-prelude/src/Hydra/Prelude.hs:
```diff
@@ -142,3 +146,8 @@ genericArbitrary ::
   Random.genericArbitrary Random.uniform
+
+-- | A seeded, deterministic, generator
+generateWith :: Gen a -> Int -> a
```
:+1:

ch1bo commented on hydra-node/src/Hydra/Ledger.hs:
```diff
@@ -7,45 +7,54 @@ import Hydra.Prelude
-  { applyTransactions :: Utxo tx -> [tx] -> Either ValidationError (Utxo tx)
-  , initUtxo :: Utxo tx
+  { -- | Apply a set of transaction to a given UTXO set. Returns the new UTXO or
+    -- validation failures returned from the ledger.
```
:heart: For the haddocks.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -558,7 +573,10 @@ stateQueryClient tracer tipVar utxoVar address =
+  ( Cardano.Api.PaymentVerificationKey (Ledger.VKey vk)
+    , Cardano.Api.PaymentSigningKey sk
+    ) <-
+    generate genKeyPair
```
Wtf fourmolu (?)

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -183,11 +194,15 @@ withTinyWallet tracer magic (vk, sk) iocp addr action = do
+                fromLedgerTxId txid
+                  `signWith` ( Cardano.Api.PaymentVerificationKey (Ledger.VKey vk)
+                             , Cardano.Api.PaymentSigningKey sk
+                             )
```
Should: Not require a keypair, but only a signing key to `signWith`.
The current signature allows to provide different vk than sk and is more a footgun than just taking the sk and deriving the vk from it.

KtorZ replied:
Indeed.

ch1bo commented on hydra-node/hydra-node.cabal:
```diff
@@ -129,6 +130,7 @@ library
     , contra-tracer
     , data-default
     , filepath
+    , formatting
```
Can you give a quick explanation / intro about this package? Never used it and also would be interested what for we use it now.

KtorZ replied:
Used to better format CBOR deserialisation error. The cardano-base package provides `Buildable` instance for the `DecoderError` which are nicer than the default `Show`... ideally, I wish `cardano-base` exposed a `prettyDecoderError` function so we wouldn't have to import this dependency.

This is debatable, we can always fallback to `show`.

ch1bo reviewed (commented):
I like your refactoring a lot. Makes everything cleaner!

Some "Shoulds" though and it's still not compiling. The PR is still Draft, but I'd like to merge this soon :)

KtorZ committed:
Add extra property to check roundtrip between cardano-api / cardano-ledger-specs.

KtorZ committed:
Continue updating to cardano-api in local-cluster.

ch1bo committed:
Use to- and fromLedgerUtxo in Hydra.Ledger.Cardano

abailly committed:
Provide missing FromJSONKey instance for API's TxIn

KtorZ committed:
Implement fromLedgerUtxo / toLedgerUtxo

KtorZ committed:
Continue fixing local-cluster.

ch1bo committed:
Make local-cluster compile

ch1bo committed:
[wip] Start fixing hydra-tui

abailly committed:
[wip] Fix more compilation errors in TUI

ch1bo committed:
[wip] Fixing some more errors in TUI

ch1bo committed:
Change mkVkAddress to take network magic

ch1bo committed:
Define utxoMap and use that in TUI

Instead of using the Data constructor and it's also a nicer name than
unUtxo

ch1bo committed:
Make TUI compile.. at a cost

The Ord AddressInEra instance is not cool and we should find another
solution.

ch1bo committed:
Use the serialized Address as keys in TUI

ch1bo committed:
Define orphan Ord for addresses using Show

KtorZ committed:
Start (re)implementing 'mkSimpleCardanoTx'

abailly committed:
Complete construction of mkSimpleTx TxBody

ch1bo committed:
Sign transaction in mkSimpleCardanoTx and changed signature

KtorZ committed:
Use Ledger's Mary generator for UTXO, as it was before.

KtorZ committed:
Use Latin1 encoding to decode Base16-encoded assetname to match cardano-api.

  Weird.

abailly committed:
Add roundtrip test for AssetName

Getting issues with invalid UTF-8 encoding of AssetName from
cardano-api's instances, which are wrong.

ch1bo committed:
Update cardano-node et al to recent master

ch1bo committed:
Fix Hydra.Network.Ouroboros with changed API

ch1bo committed:
Revive FromJSON Ledger.Mary.AssetName orphans

ch1bo committed:
Fix evaluateTransactionExecutionUnits due to changed API

ch1bo committed:
Add new golden files

ch1bo committed:
Define Arbitrary instances for TxIn and TxOut and fix commitTx TxSpec test

Both are using a generator from the ledger and the cardano-api
conversion functions.

KtorZ committed:
Write extra property for checking equality of transaction id after JSON roundtrip

  Interestingly, checking for equality of the transaction does not
  suffice because of the internal caching / memoization going on in the
  ledger; it is possible to end with two transaction values that are
  equal, but lead to different binary encoding (and thus, different
  ids). This is arguably a serialization bug in the ledger which can
  likely not be fixed at this stage without breaking mainnet.

  Whoops.

KtorZ committed:
Replace Mary generators by Alonzo generators for transactions and utxos.

  Still, not generating transactions with scripts (in principle?)

KtorZ committed:
Write JSON instances for redeemers and datums, and remove TODOs from TxWitness instances.

KtorZ committed:
Re-generate golden tests.

KtorZ committed:
re-generate materialized files.

KtorZ committed:
Drop additional constraints on generator constants

  The constants and the generator are inter-dependent and it seems that we can't change the constants freely without causing bad behavior in the generators.

KtorZ committed:
Report failing transaction with validation errors when applying to ledger

  Better debugging, in particular in tests.

KtorZ committed:
Add non-zero default max execution units for hard-coded ledger env.

ch1bo committed:
Change hard-coded ledgerEnv to use freeCostModel

The txs from `genTx` would hit execution budgets as the template scripts
used in the ledger generator were using a `freeCostModel` themselves.

ch1bo committed:
Re-generate golden/Tx

abailly committed:
Replace invalid sample with property check on non-constant dataset

KtorZ committed:
Continue fixing errors after rebase on latest master.

ch1bo committed:
Use cardano-api types in commitTx

This fixes the datum encoded into commit transactions (for now).

ch1bo committed:
Create and send a Alonzo transaction in e2e test

We now use the blessed 'Era' of Hydra.Ledger.Cardano, with the small
workaround on some redundant fields on the witnesses.

KtorZ committed:
Favor JSON instances from cardano-api when possible / available

  The instance scaffolding going on at the moment feels however a bit
  clunky / fragile.

  In the medium / long run, I think the right way of fixing this would be
  to provide those instances upstream in `cardano-api`, with "smart"
  encoders that omit null and empty values as a bonus.

KtorZ committed:
Do not encode empty or null fields ToJSON by default

  This creates needlessly large results which can be quite annoying to work with client-side.

KtorZ committed:
Add back 'id' to the JSON serialized output of transactions.

  This is required by the benchmarks and actually quite convenient for consumers.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Orphans.hs:
```diff
@@ -0,0 +1,563 @@
+  toJSON b =
+    object $
+      mconcat
+        [ onlyIf (const True) "inputs" (Set.map fromShelleyTxIn (Ledger.Alonzo.inputs' b))
```
Why the `onlyIf`, if this is always included?

KtorZ replied:
Consistency with the rest. It reads better.

ch1bo commented on hydra-node/src/Hydra/Ledger/Cardano/Orphans.hs:
```diff
@@ -0,0 +1,563 @@
+        , onlyIf ((/=) mempty) "mint" (fromMaryValue (Ledger.Alonzo.mint' b))
+        , onlyIf isSJust "scriptIntegrityHash" (Ledger.Alonzo.scriptIntegrityHash' b)
+        , onlyIf isSJust "auxiliaryDataHash" (Ledger.Alonzo.adHash' b)
+        , onlyIf isSJust "networkId" (Ledger.Alonzo.txnetworkid' b)
```
Flipping the predicate seems to be the more natural case to identify, i.e. when to exclude: `excludeIf null "requiredSignatures" .. `

ch1bo reviewed (commented):
Just saw this, no need to change now, but if you want to still comment on it, I can change it around if you agree.
